og_migrate/migrations/d7_og_entity_membership.yml

```diff
@@ -5,8 +5,26 @@ migration_tags:
 source:
   plugin: d7_og_membership
 process:
-  target_id: gid
-  entity_id: etid
+  # Currently only supports node groups
```

Comment:
We can't use migration_lookup because we have to support multiple entity types.

Reply:
Then I believe we need to write a special process plugin to lookup the correct entity.
Not using any lookup will likely not work correctly with sites that already have content, as the destination id might be already used by an existing entity.

It seemed more secure to me to only support node entities for now, rather than having memberships affected to wrong entities. No ?

Reply:
I was at a bof earlier today and about 20 minutes after I wrote this, heddn mentioned that we might be able to do something with multiple migrations in migrate_lookup or may be refactor this to be a derivative. Then we could use migration_lookup and dynamically find the migration based on the entity_type. Maybe.

Reply:
What do you mean by a derivative? A subclass of the migration_lookup plugin

I had tried a solution that involves the creation of an intermediate process that maps the entity_type to a migration plugin to use for the lookup.
Something in the lines of:

<pre>
...
  entity_lookup_plugin:
      plugin: static_map
      source: entity_type
      map:
        node: d7_node
        user: d7_user
        comment: d7_comment
  entity_id:
    -
      plugin: migration_lookup
      migration: @entity_lookup_plugin
      source: etid
    -
      plugin: skip_on_empty
      method: row
      message: 'Membership entity is missing'
...
</pre>

But I couldn't quite get it to work, as the "@entity_lookup_plugin" was replaced by "{}" when I ran `drush migrate-upgrade`, and couldn't figure out why, so I abandoned it!

Reply:
I'm not too familiar with it, but the term I was looking for was "deriver". Something like [D7NodeDeriver](https://api.drupal.org/api/drupal/core%21modules%21node%21src%21Plugin%21migrate%21D7NodeDeriver.php/class/D7NodeDeriver/8.7.x). It is not really documented too well on [Writing migrations for contributed and custom modules](https://www.drupal.org/docs/8/api/migrate-api/writing-migrations-for-contributed-and-custom-modules#node-deriver) as it just mentions that reading the source is the best way to understand it. Derivers are complex and I find it difficult to understand the concept when it comes to the migrate API.

Reply:
So I think we need to change all the d7 group migrations to be a deriver per entity_type, that can feed the correct info in, and then I think we can do a migration_lookup appropriately for it. Maybe. It's complex :|

Reply:
Okay, I think I have a plan.

1. We know that there will be potentially a matrix of x * y of migrations (i.e. group_type * entity_type).
2. We write a deriver to populate derivative migrations for this matrix based on d7_og_entity_membership.
3. We comment d7_og_entity_membership to show how someone can create a group_type * entity_type yaml for each of their group_types that aren't supported or for group_types that have different entity_type's supported.

I don't think it's possible to do anything else. I haven't found a way to look up an arbitrary migration id based on an arbitrary entity type plugin id.